src/java/org/apache/cassandra/thrift/ThriftSessionManager.java

```diff
@@ -82,4 +97,21 @@ public int getConnectedClients()
     {
         return activeSocketSessions.size();
     }
+
+    /**
+     * Currently only implemented for synchronous thrift server. This ensures that when the server socket is closed,
+     * spawned client sockets will also be closed instead of waiting for TCP timeouts.
+     */
+    public void closeActiveClients() throws IOException
```

Comment:
I might be missing some details here, so I was hoping you can provide some additional context.

My understanding is that the maximum number of active threads can be [quite large](https://github.com/palantir/cassandra/blob/palantir-cassandra-2.2.18/src/java/org/apache/cassandra/thrift/CustomTThreadPoolServer.java#L110). Can we potentially run into the following race:

1. The maximum number of possible threads are scheduled.
2. JVM schedule each thread one at a time, but before they register with `ThriftSessionManager`, they are context switched.
3. The client signals `stop`, and we remove all "active clients", but they haven't been registered in (2).
4. Scheduler context switched to the threads in (2) and they start a connection with the remote client.

My questions:
* Should we have a lock on `activeClients`?
* I'm not sure where exactly in the code we instantiate connection with the remote client, so maybe locking is not necessary.
* Additionally, does `client.close()` forcefully close the connection on the other side?

Reply:
I think most of that is possible. My opinion however is at the exact moment "closeActiveClients" is called those other non-registered connections aren't yet "active clients" (at least w/ my read of how these classes implicitly define active) b/c despite having a connection they aren't yet being processed. 

If you called stop, then another thread registered + tracked a client for an earlier created socket, line 218 in CustomTThreadPoolServer should prevent it from doing any work and then have the socket closed shortly after anyway